internal/resources/tanzukubernetescluster/resource_tanzu_kuberenetes_cluster.go

```diff
@@ -0,0 +1,424 @@
+			clusterFn.ManagementClusterName, clusterFn.ProvisionerName, clusterFn.Name)
+	}
+
+	return []*schema.ResourceData{data}, nil
+}
+
+func validateSchema(ctx context.Context, data *schema.ResourceDiff, value interface{}) error {
+	config := value.(authctx.TanzuContext)
+	topologyData := data.Get(SpecKey).([]interface{})[0].(map[string]interface{})[TopologyKey].([]interface{})[0].(map[string]interface{})
+
+	clusterClass, clusterClassExist := topologyData[ClusterClassKey].(string)
+
+	if !clusterClassExist {
+		return errors.New("Cluster Class must be set during creation.")
+	}
+
+	clusterClassFn := &clusterclassmodels.VmwareTanzuManageV1alpha1ManagementClusterProvisionerClusterClassFullName{
+		ManagementClusterName: data.Get(ManagementClusterNameKey).(string),
+		ProvisionerName:       data.Get(ProvisionerNameKey).(string),
+		Name:                  clusterClass,
+	}
+
+	resp, err := config.TMCConnection.ClusterClassResourceService.ClusterClassResourceServiceGet(clusterClassFn)
+
+	if err != nil || len(resp.ClusterClasses) == 0 {
+		return errors.Wrapf(err, "Couldn't find cluster class - Management Cluster Name: %s, Provisioner: %s, Cluster Class Name: %s.",
+			clusterClassFn.ManagementClusterName, clusterClassFn.ProvisionerName, clusterClassFn.Name)
+	}
+
+	clusterClassValidator := NewClusterClassValidator(resp.ClusterClasses[0].Spec)
+	clusterVariablesErrs := clusterClassValidator.ValidateClusterVariables(topologyData[ClusterVariablesKey].(string), true)
+
+	if len(clusterVariablesErrs) > 0 {
+		errStr := "Cluster Variables validation failed:\n"
+
+		for _, e := range clusterVariablesErrs {
+			errStr = fmt.Sprintf("%s%s\n", errStr, e.Error())
+		}
+
+		err = errors.New(errStr)
+	}
+
+	nodePoolsErrs := clusterClassValidator.ValidateNodePools(topologyData[NodePoolKey].([]interface{}))
+
+	if len(nodePoolsErrs) > 0 {
+		errStr := "Node pools validation failed:\n"
+
+		if err != nil {
+			errStr = fmt.Sprintf("%s%s", err.Error(), errStr)
+		}
+
+		for _, e := range nodePoolsErrs {
+			errStr = fmt.Sprintf("%s%s\n", errStr, e.Error())
+		}
+
+		err = errors.New(errStr)
+	}
+
+	return err
+}
+
+// Helper functions.
+
+func readResourceWait(ctx context.Context, config *authctx.TanzuContext, resourceFullName *tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1ManagementClusterProvisionerTanzukubernetesClusterFullName) (resp *tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1ManagementClusterProvisionerTanzukubernetesClusterData, err error) {
+	stopStatuses := map[tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1CommonClusterStatusPhase]bool{
+		tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1CommonClusterStatusPhaseERROR:         true,
+		tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1CommonClusterStatusPhaseUPGRADEFAILED: true,
+		tanzukubernetesclustermodels.VmwareTanzuManageV1alpha1CommonClusterStatusPhaseREADY:         true,
```

Comment:
@GilTeraSky - We discussed about this on Friday and today we had an internal discussion so based on that here are the changes we would request you to do.

Schema should include the following
1. By default as per your implementation the operations would wait on these stop statuses. A minor change I would request you is to look for status phase `VmwareTanzuManageV1alpha1CommonClusterStatusPhaseHEALTHY` rather than `VmwareTanzuManageV1alpha1CommonClusterStatusPhaseREADY`. But in some unexpected cases if at all this can lead to infinite wait loop, lets ensure that we have  wait_timeout field in the schema with a  default timeout (if not specified by the user with which we can terminate the wait loop in worst scenarios, like the one you observed with TKGm delete in acceptance tests).
2. Add 2 more fields to the schema a bool `waitForKubeConfig` and `kubeconfig`(computed field which is for printing the kubeconfig as output). The second field would be set only when the `waitForKubeConfig` is true.
For reference you can see this  [PR](https://github.com/vmware/terraform-provider-tanzu-mission-control/pull/313/). This one has been an ask from our terraform users , so lets try to address this

cc: @vmw-vjn

Reply:
Thanks @ramya-bangera . Only thing I'll add (may be obvious, so sorry if its repetitive) is that:
said wait_timeout field value (whether it be default or user-provided), should be the main counter to terminate all/any wait loops - including for `VmwareTanzuManageV1alpha1CommonClusterStatusPhaseHEALTHY ` as well as `waitForKubeConfig `

Reply:
Just to clarify - if the timeout is expired, should I fail the resource?

Reply:
If timeout is expired, then I'd say the should exit based on resource (cluster) state.

- Cluster is in healthy state at timeout but pinniped is not healthy. Exit with resource creation success but kubeconfig output set as null; assuming that `waitforKubeConfig` flag has been set by user. Optionally a warning or message stating that timeout occurred and KubeConfig couldn't be retrieved in set time (for example).
- Cluster is in ready state, not yet healthy at timeout. Exit with resource creation success, but warning or message stating resource was created but may not be healthy (for example).
- Cluster is still processing state, not yet ready at timeout. Exit with resource creation fail and rollback/clean-up where possible. Warning or message stating resource could not be successfully created in the set time (for example)

@ramya-bangera - please comment if I've got something incorrect or missed a scenario in list

Reply:
As agreed today in Zoom call, I believe the path forward is to implement another flag for setting timeout behavior. If set to true (default), on time-out the resource creation/change  loop is exited and resource creation is failed in TF and if set to false, on time-out the resource creations loop is exited and resource creation is not failed in TF, if cluster is still in processing or ready or deploying stages.